### ISSUE TITLE
Does flexmark-java provide OSGi support?

### ISSUE THREAD

**Issue by ieugen** [vsch/flexmark-java#236]
Hello,

I would like to useflexmark-java to render markdown content inside an Apache Sling application. 
Does this library provide OSGi support? If not, are you considering it?

Thanks,

**Comment by vsch** [comment]
@ieugen, I have not considered it. I only took a quick look at OSGi site. 

What would be the advantage of using OSGi and why can't the library be used in an OSGi module as it is without the library needing to address OSGi modularity?

**Comment by ieugen** [comment]
Well, in short, the advantage will be that the application will be available to more users. 
Also, if the code works in OSGi it is a good indication that it is well designed (OSGi kind of forces you to separate API and implementation - or at least think about it). 

To support OSGi, you will need to provide bundles. A bundle = jar + special MANIFEST.MF. 
This is very easily generated in maven using http://felix.apache.org/components/bundle-plugin/ . 
See http://felix.apache.org/documentation/subprojects/apache-felix-maven-bundle-plugin-bnd.html for a simple example (http://felix.apache.org/documentation/subprojects/apache-felix-maven-bundle-plugin-bnd.html#using-the-plugin). 

Most of the time it is enough just to add and configure the plugin and change packaging to `bundle` and that should be enaough.

TLDR: 

OSGi is a specification + framework on top of Java/JVM for writing modular applications. It forces modularity on a package basis. On top of that it provides a dynamic Service layer + compendium of services. It's quite awesome but requires some learning curve. 

OSGi is used by Eclipse IDE and also Netbeans IDE if I remember correctly. I'm planning to use it with Apache Sling deployed on top of Apache Karaf application container.

**Comment by vsch** [comment]
@ieugen, thank you for the information.

I would point out that the advantage of OSGi for this library would be limited as the library modules are highly coupled to the core version and need to be used together. Dynamic loading and modularity is not very critical.

I am also skeptical of anything being able to force code to be well designed. Niklaus Wirth thought that a structured language like Pascal would achieve that, until he was shown a program by some proud programmer written in Pascal, and Niklaus said it was the worst piece of convoluted spaghetti he ever saw. He realized that his idealist view was wrong.

As far as being accepted goes, that too is ephemeral to me. The library was a solution to a problem of performance and complexity of adding extensions to pegdown. Beyond solving real problems for real use cases, I am not motivated to fiddle with it. If someone finds it useful, great. If someone wants to contribute and help improve it, even better.  Do I want to put effort into making it popular for the sake of popularity? Not at all.

Of course it can be improved or done better. Any software can. I firmly believe that software is like a poem. It is never finished. It is merely abandoned.

However, I only have 24 hours in a day and need at least 6 for non-software development related activity. ;)

**Comment by ieugen** [comment]
Hi, I completely understand. OSGi has ways of making coupled things work together: bundle them all into a super jar :). My sentence with "OSGi forces" was a bit exaggerated. 

Anyway, I understand that you are not willing to invest time in something that is not productive to you. Are you willing to accept PR's ? I may take a crack at it and provide minimal OSGi configuration.

**Comment by vsch** [comment]
@ieugen, I am willing to accept PRs and wish there were more of them. If it improves the library, whether for my use of someone else's I am all for it.

As long as the OSGi does not force it to be used only in an OSGi aware application I have no issues with it.

**Comment by ieugen** [comment]
Thanks. I'll see if I can make some time for that.

**Comment by vsch** [comment]
@ieugen, OSGi module support thanks to [@klcodanr](https://github.com/klcodanr) was added in 0.34.22
